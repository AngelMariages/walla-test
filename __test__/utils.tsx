import { render } from '@testing-library/react';
import {
    FavoritesContext,
    FavoritesContextType,
} from 'context/FavoritesContext';
import { FiltersContext, FiltersContextType } from 'context/FiltersContext';

type ContextValues = {
    filtersContext?: Partial<FiltersContextType>;
    favoritesContext?: Partial<FavoritesContextType>;
};

export const renderWithContext = (
    ui: React.ReactElement,
    contextValues?: ContextValues
) => {
    return render(
        <FiltersContext.Provider
            value={{
                search: '',
                setSearch: jest.fn(),
                page: 1,
                incrementPage: jest.fn(),
                sort: {
                    sortBy: 'title',
                    sortOrder: 'asc',
                },
                setSort: jest.fn(),
                ...(contextValues?.filtersContext || {}),
            }}>
            <FavoritesContext.Provider
                value={{
                    favorites: [],
                    favoritesVisible: false,
                    addFavorite: jest.fn(),
                    removeFavorite: jest.fn(),
                    isFavorite: jest.fn(),
                    toggleFavorites: jest.fn(),
                    ...(contextValues?.favoritesContext || {}),
                }}>
                {ui}
            </FavoritesContext.Provider>
        </FiltersContext.Provider>
    );
};
