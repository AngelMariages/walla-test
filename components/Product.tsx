import { forwardRef } from 'react';
import { Item } from '../pages/api/items';

type Props = {
    item: Item;
};

const Product = forwardRef<HTMLDivElement, Props>(({ item }, ref) => {
    return (
        <div className="m-5 border bg-slate-50 rounded-2xl" ref={ref}>
            <img src={item.image} width={150} alt={item.title} />
            <p className='text-lg font-bold'>{item.title}</p>
            <p>{item.price}</p>
            <p>{item.description}</p>
        </div>
    );
});

export default Product;
