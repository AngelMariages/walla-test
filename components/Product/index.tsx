import Image from 'next/image';
import { forwardRef, useContext } from 'react';
import FavIcon from 'public/icons/fav.svg';
import { FavoritesContext } from 'context/FavoritesContext';
import { Item } from 'constants/item';

type Props = {
    item: Item;
    selected?: boolean;
};

const formatPrice = (price: number) => {
    return new Intl.NumberFormat('es-ES', {
        style: 'currency',
        currency: 'EUR',
        minimumFractionDigits: 0,
    }).format(price);
};

export default forwardRef<HTMLDivElement, Props>(function Product(
    { item, selected = false },
    ref
) {
    const { addFavorite, removeFavorite } = useContext(FavoritesContext);

    return (
        <div
            className="m-5 p-5 border bg-slate-50 rounded-2xl shadow-2xl"
            ref={ref}>
            <div className="w-full h-80 mb-4 relative overflow-hidden rounded-2xl">
                <Image
                    src={item.image}
                    alt={item.title}
                    className="md:hover:scale-125 transition-transform duration-500 ease-out"
                    fill
                    style={{
                        objectFit: 'cover',
                    }}
                />
                <FavIcon
                    data-testid="favicon"
                    onClick={() => {
                        if (selected) {
                            removeFavorite(item);
                        } else {
                            addFavorite(item);
                        }
                    }}
                    className={`bg-white p-2 rounded-full absolute top-4 right-4 w-10 duration-500 ease-[cubic-bezier(0.95,0.05,0.795,0.035)] transition-colors ${
                        selected ? 'fill-red-500' : 'fill-transparent'
                    }`}
                />
            </div>
            <p className="text-lg font-bold">{item.title}</p>
            <p
                data-testid="price"
                className="text-md font-semibold text-right my-4 bg-slate-200 rounded-full inline-block px-4 py-1">
                {formatPrice(item.price)}
            </p>
            <p>{item.description}</p>
        </div>
    );
});
