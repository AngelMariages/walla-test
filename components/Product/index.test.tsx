import { getByTestId, act } from '@testing-library/react';
import { renderWithContext } from '__test__/utils';
import Product from '.';

jest.mock('next/image', () => {
    return function NextImage({
        children,
        ...props
    }: {
        children: React.ReactNode;
    }) {
        return (
            // eslint-disable-next-line @next/next/no-img-element
            <img
                alt="test-img"
                {...{
                    ...props,
                    fill: undefined,
                }}>
                {children}
            </img>
        );
    };
});

describe('Product', () => {
    it('should render correctly', () => {
        const item = {
            title: 'title',
            price: 300,
            description: 'description',
            email: 'item@email.com',
            image: 'image.jpg',
        };

        const { container } = renderWithContext(<Product item={item} />);
        expect(container).toMatchSnapshot();
    });

    it('should render correctly with selected', () => {
        const item = {
            title: 'title',
            price: 300,
            description: 'description',
            email: 'selected@mail.com',
            image: 'image.jpg',
        };

        const { container } = renderWithContext(
            <Product item={item} selected />
        );
        expect(container).toMatchSnapshot();
    });

    it('should call addFavorite or removeFavorite depending on selected state', async () => {
        const item = {
            title: 'title',
            price: 300,
            description: 'description',
            email: 'selected@mail.com',
            image: 'image.jpg',
        };

        const addFavorite = jest.fn();
        const removeFavorite = jest.fn();

        const { container } = renderWithContext(<Product item={item} />, {
            favoritesContext: {
                addFavorite,
                removeFavorite,
            },
        });

        expect(addFavorite).not.toHaveBeenCalled();
        expect(removeFavorite).not.toHaveBeenCalled();

        const favButton = getByTestId(container, 'favicon');
        await act(() => favButton.click());

        expect(addFavorite).toHaveBeenCalled();
        expect(removeFavorite).not.toHaveBeenCalled();

        jest.resetAllMocks();

        const { container: container2 } = renderWithContext(
            <Product item={item} selected />,
            {
                favoritesContext: {
                    addFavorite,
                    removeFavorite,
                },
            }
        );

        const favButton2 = getByTestId(container2, 'favicon');
        await act(() => favButton2.click());

        expect(addFavorite).not.toHaveBeenCalled();
        expect(removeFavorite).toHaveBeenCalled();
    });

    it('should format price correctly', () => {
        const item = {
            title: 'title',
            price: 302.45,
            description: 'description',
            email: 'test@test.com',
            image: 'image.jpg',
        };

        const { container } = renderWithContext(<Product item={item} />);
        expect(container).toMatchSnapshot();

        const price = getByTestId(container, 'price');

        expect(price.textContent).toBe('302,45\xa0â‚¬');
    });
});
