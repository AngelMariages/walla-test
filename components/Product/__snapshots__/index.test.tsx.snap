// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product should format price correctly 1`] = `
<div>
  <div
    class="m-5 p-5 border bg-slate-50 rounded-2xl shadow-2xl"
  >
    <div
      class="w-full h-80 mb-4 relative overflow-hidden rounded-2xl"
    >
      <img
        alt="title"
        class="md:hover:scale-125 transition-transform duration-500 ease-out"
        src="image.jpg"
        style="object-fit: cover;"
      />
      <svg-url
        classname="bg-white p-2 rounded-full absolute top-4 right-4 w-10 duration-500 ease-[cubic-bezier(0.95,0.05,0.795,0.035)] transition-colors fill-transparent"
        data-testid="favicon"
      />
    </div>
    <p
      class="text-lg font-bold"
    >
      title
    </p>
    <p
      class="text-md font-semibold text-right my-4 bg-slate-200 rounded-full inline-block px-4 py-1"
      data-testid="price"
    >
      302,45 €
    </p>
    <p>
      description
    </p>
  </div>
</div>
`;

exports[`Product should render correctly 1`] = `
<div>
  <div
    class="m-5 p-5 border bg-slate-50 rounded-2xl shadow-2xl"
  >
    <div
      class="w-full h-80 mb-4 relative overflow-hidden rounded-2xl"
    >
      <img
        alt="title"
        class="md:hover:scale-125 transition-transform duration-500 ease-out"
        src="image.jpg"
        style="object-fit: cover;"
      />
      <svg-url
        classname="bg-white p-2 rounded-full absolute top-4 right-4 w-10 duration-500 ease-[cubic-bezier(0.95,0.05,0.795,0.035)] transition-colors fill-transparent"
        data-testid="favicon"
      />
    </div>
    <p
      class="text-lg font-bold"
    >
      title
    </p>
    <p
      class="text-md font-semibold text-right my-4 bg-slate-200 rounded-full inline-block px-4 py-1"
      data-testid="price"
    >
      300 €
    </p>
    <p>
      description
    </p>
  </div>
</div>
`;

exports[`Product should render correctly with selected 1`] = `
<div>
  <div
    class="m-5 p-5 border bg-slate-50 rounded-2xl shadow-2xl"
  >
    <div
      class="w-full h-80 mb-4 relative overflow-hidden rounded-2xl"
    >
      <img
        alt="title"
        class="md:hover:scale-125 transition-transform duration-500 ease-out"
        src="image.jpg"
        style="object-fit: cover;"
      />
      <svg-url
        classname="bg-white p-2 rounded-full absolute top-4 right-4 w-10 duration-500 ease-[cubic-bezier(0.95,0.05,0.795,0.035)] transition-colors fill-red-500"
        data-testid="favicon"
      />
    </div>
    <p
      class="text-lg font-bold"
    >
      title
    </p>
    <p
      class="text-md font-semibold text-right my-4 bg-slate-200 rounded-full inline-block px-4 py-1"
      data-testid="price"
    >
      300 €
    </p>
    <p>
      description
    </p>
  </div>
</div>
`;
